core/domain/html_validation_service.py

```diff
@@ -0,0 +1,890 @@
+                parent not in allowed_parent_list[tag.name]):
+            if tag.name in err_dict:
+                err_dict[tag.name].append(parent)
+            else:
+                err_dict[tag.name] = [parent]
+            is_invalid = True
+
+    return is_invalid
+
+
+def add_caption_attr_to_image(html_string):
+    """Adds caption attribute to all oppia-noninteractive-image tags.
+
+    Args:
+        html_string. str: HTML string in which the caption attribute is to be
+            added.
+
+    Returns:
+        str. Updated HTML string with the caption attribute for all
+            oppia-noninteractive-image tags.
+    """
+    soup = bs4.BeautifulSoup(
+        html_string.encode(encoding='utf-8'), 'html.parser')
+
+    for image in soup.findAll(name='oppia-noninteractive-image'):
+        attrs = image.attrs
+        if 'caption-with-value' not in attrs:
+            image['caption-with-value'] = escape_html(json.dumps(''))
+
+    return unicode(soup)
+
+
+def validate_customization_args(html_list):
+    """Validates customization arguments of Rich Text Components in a list of
+    html string.
+
+    Args:
+        html_list: list(str). List of html strings to be validated.
+
+    Returns:
+        dict: Dictionary of all the invalid customisation args where
+            key is a Rich Text Component and value is the invalid html string.
+    """
+    # Dictionary to hold html strings in which customization arguments
+    # are invalid.
+    err_dict = {}
+    rich_text_component_list = [
+        'oppia-noninteractive-link',
+        'oppia-noninteractive-math',
+        'oppia-noninteractive-image',
+        'oppia-noninteractive-video',
+        'oppia-noninteractive-collapsible',
+        'oppia-noninteractive-tabs'
+    ]
+
+    for html_string in html_list:
+        soup = bs4.BeautifulSoup(
+            html_string.encode(encoding='utf-8'), 'html.parser')
+
+        for tag_name in rich_text_component_list:
+            for tag in soup.findAll(name=tag_name):
+                err_msg = _validate_customization_args_in_tag(tag)
+
+                if err_msg:
+                    if err_msg not in err_dict:
+                        err_dict[err_msg] = [html_string]
+                    elif html_string not in err_dict[err_msg]:
+                        err_dict[err_msg].append(html_string)
+
+            if tag_name not in [
+                    'oppia-noninteractive-collapsible',
+                    'oppia-noninteractive-tabs']:
+                for collapsible in soup.findAll(
+                        name='oppia-noninteractive-collapsible'):
+                    # This is wrapped in try catch block since any invalid
+                    # content such as int will not be convertible in Beautiful
+                    # Soup format. Also missing attributes will result in error
+                    # when we try to extract them here.
+                    try:
+                        content_html = json.loads(
```

Comment:
I think you should just extract tags in this function. Don't do deep inspection of the tags (which in this case includes extracting HTML from within the tag). Let the second function handle this as part of the component-specific validation.

Reply:
Done!